lib/queue.js

```diff
@@ -223,6 +223,91 @@ Queue.prototype.checkHealth = function (cb) {
   return promise;
 };
 
+Queue.prototype._scanForJobs = function(cursor, set, cb) {
+  // Ideally, we would be able to paginate redis sets like we do sorted ranges.
+  // For an MVP, however, we're aiming to replace our usage of Arena for Bull 1.x
+  // and Bull 1.x calls SMEMBERS -- which is more blocking than SSCAN.
+
+  this.client.sscan(
+    this.toKey('failed'),
+    cursor,
+    'COUNT', '100',
+    (err, results) => {
+      const [nextCursor, ids] = results;
+
+      // A given element may be returned multiple times in SSCAN.
+      // So, we use a set to remove duplicates.
+      // https://redis.io/commands/scan#scan-guarantees
+      for (let id of ids) set.add(id);
+
+      if (nextCursor === '0') {
+        return cb(null, set);
+      }
+
+      this._scanForJobs(cb, nextCursor, set);
+    }
+  );
+}
+
+// TODO(randall): Implement (start, end) parameters
+Queue.prototype._getJobs = function (queueType, start, end, cb) {
+  let promise;
+
+  if (this.paused) {
+    promise = Promise.reject(new Error('closed'));
+  } else {
+    const idsPromise = helpers.deferred();
+    this._ready
+      .then(() => {
+        if (this.paused) {
+          throw new Error('closed');
+        }
+      })
+      .then(() => {
+        switch (queueType) {
+          case 'failed':
+          case 'succeeded':
+            this._scanForJobs('0', new Set(), idsPromise.defer());
+            break;
+          case 'waiting':
+          case 'active':
+            this.client.lrange(this.toKey(queueType), start, end, idsPromise.defer());
+            break;
+          case 'delayed':
+            this.client.zrange(this.toKey(queueType), start, end, idsPromise.defer());
+            break;
+        }
+      });
+
+    promise = idsPromise.then((ids) => {
+      return Promise.all([...ids].map((id) => this.getJob(id)));
+    });
+  }
+
+  if (cb) helpers.asCallback(promise, cb);
+  return promise;
+}
+
+Queue.prototype.getWaiting = function (start, end, cb) {
+  return this._getJobs('waiting', start, end, cb);
+}
+
+Queue.prototype.getActive = function (start, end, cb) {
+  return this._getJobs('active', start, end, cb);
+}
+
+Queue.prototype.getSucceeded = function (cb) {
+  return this._getJobs('succeeded', null, null, cb);
+}
+
+Queue.prototype.getFailed = function (cb) {
+  return this._getJobs('failed', null, null, cb);
+}
+
+Queue.prototype.getDelayed = function (start, end, cb) {
+  return this._getJobs('delayed', start, end, cb);
```

Comment:
Yeah, the only reason we do it like this is because that was the convention Bull used. I agree this would probably be cleaner.